Cache translations and _() macro (LOCALIZE=1 build perf improved ≈20%)

Aivean opened the pull request:
#### Summary

SUMMARY: Performance "Cache translations and _() macro (LOCALIZE=1 build perf improved ≈20%)"

#### Purpose of change

As @ZhilkinSerg pointed out, currently localized version of the game is significantly slower than non-localized one. 
Depending on the scenario, a translation call inside `mtype::get_footsteps` can take up to 12% of the game turn time (when waiting).

Other places, like `press_x` are called very frequently while waiting and `_(...)` call inside them becomes a hotspot, taking 1-2% of CPU time.

#### Describe the solution

I decided to fix current and (potentially) future performance problems with translations by doing three things:
1. reduced the size of `translation` class by moving optional members to the heap (using `cata::value_ptr`). This change alone improved game performance (both localized and non-localized version) by  ≈2%
2. added versioned cache to `translation` class, that caches the result of the last call of `translation::translated` method and is reset by language change
3. changed `_( ... )` implementation to macro that locally caches its translation (similarly to `STATIC()` macro) and is also reset by language change

Overall, these changes dropped the translation overhead to zero and, due to the reduced translation size, even improved non-localized performance somewhat.

#### Describe alternatives you've considered

Caching only the `translation` class and using it instead of `_()` macro in "hot" places. However, the problem with that approach would be that in the future people will continue to use convenient `_()` macro, potentially disregarding performance considerations. 

With the current approach, `_()` call is cached for string literals and won't introduce performance problems in the future. 

#### Testing

Ensured that both localized and non-localized version of the game compiles and runs. 
Checked that changing language in the settings actually changes the language in the UI.

#### Additional context

Before (non-localized version):
<img  width=400 src="https://user-images.githubusercontent.com/2865203/97947947-ca5c6800-1d43-11eb-95f6-aea9949dd206.png"/>

Before (localized version):
<img  width=400 src="https://user-images.githubusercontent.com/2865203/97947962-d7795700-1d43-11eb-8b6e-f29a32996e15.png"/>

After (localized version):
<img  width=400 src="https://user-images.githubusercontent.com/2865203/97947986-e8c26380-1d43-11eb-9aaa-1941e8544627.png"/>

Profiling
Before:
![image](https://user-images.githubusercontent.com/2865203/97948120-71410400-1d44-11eb-94db-f83d1797f0a7.png)

After:
![image](https://user-images.githubusercontent.com/2865203/97948095-579fbc80-1d44-11eb-9324-d8f53d8396db.png)

----

I can see only one potential drawback of this solution:  I don't think that the `ATTRIBUTE_FORMAT_ARG( 1 )` attribute of the `_( )` function works any longer, since it's converted to macro (not sure how to check).

----

@jbytheway , I hate to bother you again, but I believe this change lies within your area of expertise, so I'd really appreciate it if you could take a look.

Qrox commented on src/translations.h:
```diff
@@ -50,15 +59,45 @@
+#define _( msg ) \
+    ( ( []( const auto & arg )-> const auto & { \
+        static int _cached_lang_version = details::get_current_language_version(); \
+        static auto * _cached_arg_ptr = &arg; \
```
If `msg` is `const char *`, the inferred type of `arg` is `const char *const &`, so this cached pointer becomes the stack address of the pointer instead of the pointer itself. If `arg` is then invalidated and points to the same string at a different memory location and `_( msg )` is called again, the invalid `_cached_translation` will probably still get returned.

I suggest either doing all the operations with `std::string` to avoid the issue altogether, or adding some unit test to ensure this macro is acting correctly.

Also, if `msg` is `std::string` and is at a different address when this macro is called again (when `msg` is on the heap, for example), comparing the pointer would completely disable the cache and degrade the performance.

Aivean replied:
Thanks for the review.
You are right regarding the stack address, although I think `_cached_translation` won't be returned in the described scenario, I think there might be safer ways to implement this.

Qrox commented:
Did you compile a `.mo` file for English before testing the performance? Because there was a libintl bug that caused low performance when a `.mo` file is not found, and I'd like to see if this change can work around that bug (and how much performance improvement can we still get if we already work around it by compiling a `.mo` file for English).

Also, it would also be nice if you test the performance in some other languages (the performance might be different because of string look-up in libintl).

Aivean commented:
> Did you compile a `.mo` file for English before testing the performance? Because there was a libintl bug that caused low performance when a `.mo` file is not found, and I'd like to see if this change can work around that bug (and how much performance improvement can we still get if we already work around it by compiling a `.mo` file for English).

I tested using `LANGUAGE=all`, so there was `.mo` file for english. Can you please suggest the specific `make` flags that would produce the scenario you want to evaluate?
Also, please note, I'm working on mac, while the bug, that you mentioned, seems to be specific to MinGW (according to the docs).

> It would also be nice if you test the performance in some other languages (the performance might be different because of string look-up in libintl).

I can do that, but my profiling shows that in the waiting scenario lookups are avoided altogether.

Aivean commented:
Hmm, seems that converting `_()` to macro put clang-tidy off: 
```
Translator comment without a matching raw string [cata-translator-comments,-warnings-as-errors]
```

(although it probably won't change the behavior of `update_pot.sh`)

This definitely needs more work, marking as a draft for now.

Qrox commented:
> I tested using LANGUAGE=all, so there was .mo file for english. Can you please suggest the specific make flags that would produce the scenario you want to evaluate?
Also, please note, I'm working on mac, while the bug, that you mentioned, seems to be specific to MinGW (according to the docs).

I see, so the 20% perf improvement is not related to the libintl bug, that's very nice.

> Hmm, seems that converting _() to macro put clang-tidy off:

The clang-tidy check currently assumes that `_()` is a function, so the check should be updated to also check for macros.

Aivean commented:
@Qrox , I fixed the `_()` cache implementation and added some tests.

Internally, there are two specializations for the `_()` cache:  for `char *` and `string` argument.  
String variant is rather straightforward. Note, it returns the result by value (from the macro) for safety.

`char *` variant is a bit more involved. It explicitly checks if the returned translation points to the same address as the argument and if yes, it's assumed that translation is absent and, for the same **value** of the argument, currently passed argument will be returned. If translation is present, it is cached and returned as usual (its address should be stable as per `gettext` docs).

Aivean commented:
Latest performance results (same savefile as in the description):

en:
![image](https://user-images.githubusercontent.com/2865203/98052485-32b25480-1deb-11eb-9e24-119210a7f7d3.png)

ru:
![image](https://user-images.githubusercontent.com/2865203/98052500-38a83580-1deb-11eb-9677-4a47b5a22b3c.png)

ZhilkinSerg commented on tests/translations_test.cpp:
```diff
@@ -0,0 +1,86 @@
+    // should return a stable address of translation that is different from the argument
+    CHECK( TRANSLATE_MACRO( "test_string" ) != test_string );
+}
+#endif
```
```suggestion
#endif

```

jbytheway commented on src/translations.cpp:
```diff
@@ -52,6 +52,15 @@ static void reload_names()
 
 static bool sanity_checked_genders = false;
 
+// int version/generation that is incremented each time language is changed
+// used to invalidate translation cache
+static int current_language_version = INVALID_LANGUAGE_VERSION + 1;
+
+int details::get_current_language_version()
+{
+    return current_language_version;
+}
+
```
Since we're using this versioned cache idea in a couple of places now it might be worth considering creating a common abstraction?  Not sure how easy that would be.

Aivean replied:
Good idea, I'll think about it. However, I believe it's outside of the scope of this PR.

jbytheway commented on src/translations.h:
```diff
@@ -12,6 +12,15 @@
+
+constexpr int INVALID_LANGUAGE_VERSION = 0;
+
+namespace details
```
I feel that `details` namespaces should be named with the thing they are the details of, otherwise they're not really providing much namespacing (I realise that they do still provide a documentary purpose).  So I'd suggest `translation_details` here.

Aivean replied:
My intention of using that namespace was to highlight that it is [an implementation internals/details](https://stackoverflow.com/questions/26546265/what-is-the-detail-namespace-commonly-used-for).  Probably should rename it to `detail` to match the name used in `assign.h` and `cata_io.h`.

Aivean replied:
Renamed the namespace to "detail".

jbytheway commented on src/translations.h:
```diff
@@ -50,16 +59,74 @@
+
+// Note: in case of std::string argument, the result is copied, this is intended (for safety)
+#define _( msg ) \
+    ( ( []( const auto & arg ) -> auto { \
```
```suggestion
    ( ( []( const auto & arg ) { \
```
The `-> auto` is redundant, I think, but you can leave it if you think it adds clarity.

Aivean replied:
You're right, I believe the comment is sufficient.

Qrox commented on tests/translations_test.cpp:
```diff
@@ -0,0 +1,86 @@
+    const char *test_string = "thread";
+
+    // should return a stable address of translation that is different from the argument
+    CHECK( TRANSLATE_MACRO( "test_string" ) != test_string );
```
```suggestion
    CHECK( TRANSLATE_MACRO( test_string ) != test_string );
```
Maybe this is what you intended?
Also I think this test will only pass if a `.mo` file for English is present.

Aivean replied:
You're right in both cases.
I fixed the test and verified that it works locally, but disabled it by default.

Aivean commented:
@Qrox , I'd greatly appreciate any advice on how to deal with the clang-tidy `cata-translator-comments` error. I haven't written clang-tidy plugins before; looking at `TranslatorCommentsCheck.cpp` and it feels like a completely unfamiliar DSL.

Qrox commented:
I think you can add a new if branch to `TranslatorCommentsCheck.cpp` at line 154 to test `_()` along with `translate_marker` and `translate_marker_context`. You also need to update the subsequent code to distinguish between `_()` and translate markers.

Aivean commented:
> I think you can add a new if branch to `TranslatorCommentsCheck.cpp` at line 154 to test `_()` along with `translate_marker` and `translate_marker_context`. You also need to update the subsequent code to distinguish between `_()` and translate markers.

Thanks, I'll try that.

Aivean commented:
I think I found an easier way to fix clang-tidy. I wrapped the argument of the expanded lambda in a superfluous marker function call and added that function name to the clang-tidy plugin. Seems to work well enough.

Aivean commented:
Ok, all issues are fixed, all suggestions resolved. I believe this is good to go.

Qrox commented:
> It probably won't work "as is", because _() expands into expression where string literal is not in the beginning (different from all other "marker" macro) but I get the idea and can work from here.

The clang-tidy check does not check whether the string literal is at the beginning of a translation call. Instead it marks any translator comments directly preceding the string literal and reports any unmarked translator comments. Since in your current fix the string literal is also not at the beginning of the expanded expression, if your solution works and mine did not work it has to be because the string literal did not get matched.

Also please update the clang-tidy test `tools/clang-tidy-plugin/test/translator-comments.cpp` because `_()` is still a function there (it has to be redeclared in the test because clang-tidy check uses the `-nostdinc++` flag)

Aivean commented:
@Qrox,

> > It probably won't work "as is", because _() expands into expression where string literal is not in the beginning (different from all other "marker" macro) but I get the idea and can work from here.
> 
> The clang-tidy check does not check whether the string literal is at the beginning of a translation call. Instead it marks any translator comments directly preceding the string literal and reports any unmarked translator comments. Since in your current fix the string literal is also not at the beginning of the expanded expression, if your solution works and mine did not work it has to be because the string literal did not get matched.

Sorry for the confusion.

When I wrote that it "probably won't work" I was thinking about macro expansion affecting `Check.MarkedStrings` location. But, to be frank, I don't know how exactly `Args->getUnexpArgument` works in with macro expansion. I haven't finished implementing your solution (so I can't say for sure whether it works or not), I came up with an easier (current) implementation in the middle of the process. 


> Also please update the clang-tidy test `tools/clang-tidy-plugin/test/translator-comments.cpp` because `_()` is still a function there (it has to be redeclared in the test because clang-tidy check uses the `-nostdinc++` flag)

I'll make the follow up PR with the tests ASAP.

Qrox commented:
Thanks for the clarification. I just tested my solution and it worked, so I'll make a PR along with the updated unit test.
